docs/apps/rbac.md

```diff
@@ -0,0 +1,538 @@
+
+Combining these functions, a hypothetical app setup might look like:
+
+```python
+from my_app.models import MyModel
+
+from rest_framework import viewsets
+from rest_framework import serializers
+
+from ansible_base.rbac.api.related import RelatedAccessMixin
+from ansible_base.rbac.api.permissions import AnsibleBaseObjectPermissions
+
+
+class MyModelSerializer(RelatedAccessMixin, serializers.ModelSerializer):  # (1)
+    class Meta:
+        model = MyModel
+        fields = '__all__'
+
+
+class MyModelViewSet(viewsets.ModelViewSet):
+    permission_classes = [AnsibleBaseObjectPermissions]  # (2)
+    serializer_class = MyModelSerializer
+
+    def get_queryset(self):
+        return MyModel.access_qs(self.request.user)  # (3)
+```
+
+This marks 3 different integration points.
+1. `RelatedAccessMixin` - checks access to related objects and gives access to users who create new objects
+2. `AnsibleBaseObjectPermissions` - checks object permission for detail views and global "add" permission
+3. `MyModel.access_qs` - filters the queryset to only objects the request user has permission to view
+
+### Creating a New Role Definition
+
+Developers should interact with `RoleDefinition` as an ordinary model.
+Giving and removing permissions should be handled by methods on a `RoleDefinition` instance.
+Ordinarily, the other models from this app should not be interacted with.
+
+NOTE: At various times a "role definition" may be referred to as just a "role".
+
+Roles are expected to be user-defined in many apps. Example of creating a custom role definition:
+
+```
+from awx.ansible_base.models.rbac import RoleDefinition
+
+rd = RoleDefinition.objects.get_or_create(name='JT-execute', permissions=['execute_jobtemplate', 'view_jobtemplate'])
+```
+
+### Assigning Permissions
+
+With a role definition object like `rd` above, you can then give out permissions to objects.
+ - `rd.give_permission(user, obj)` - give permissions listed in `rd` to that user for just that object
+ - `rd.remove_permission(user, obj)` - remove permission granted by only that role
+ - `rd.give_global_permission(user)` - if configured, give user execute/view permissions to all job templates in the system
+ - `rd.remove_global_permission(user)`
+
+These cases assume an `obj` of a model type tracked by the RBAC system,
+All the `give_permission` method will return a `RoleUserAssignment` or `RoleTeamAssignment` object.
+The `remove_permission` method will return nothing.
+Assignments have an associated `object_role` in case you need that.
+Removing permission will delete the object role if no other assignments exist.
+
+### Registering Models
+
+Any Django Model (except your user model) can
+be made into a resource in the RBAC system by registering that resource in the registry.
+
+```
+from awx.ansible_base.utils.permission_registry import permission_registry
+
+permission_registry.register(MyModel, parent_field_name='organization')
```

Comment:
Should this be done in init or just anywhere in the model's file or somewhere else?

Reply:
In all 3 concrete examples we have, it was the `models/__init__.py` or just `models.py`. But I don't want to express any particular recommendation for it. The hard requirement would be sometime before apps are ready, but this is hardly a requirement at all.